### ISSUE TITLE
NumberFormatException with some input values

### ISSUE THREAD

**Issue by ticofab** [oschrenk/sunrise.scala#2]
When calling `SunriseSunset.of` from the `0.2.0` version with the following values

```scala
latitude = 71.8333
longitude = 102.4833
date = {LocalDate@3365} "2017-07-27"
zoneId = {ZoneRegion@3366} "Asia/Krasnoyarsk"
```

The library throws an exception: 
```scala
java.lang.NumberFormatException was thrown.
java.lang.NumberFormatException
	at java.math.BigDecimal.<init>(BigDecimal.java:494)
	at java.math.BigDecimal.<init>(BigDecimal.java:824)
	at scala.math.BigDecimal$.decimal(BigDecimal.scala:51)
	at scala.math.BigDecimal$.decimal(BigDecimal.scala:54)
	at scala.math.BigDecimal$.double2bigDecimal(BigDecimal.scala:342)
	at com.oschrenk.sunrise.SunriseSunset$.of(SunriseSunset.scala:78)
```

This happens here:

```scala
val hourAngle = Math.acos(
      (sinOfAltitudeSolarDisc - sin(latitude.toRadians) * sin(sunDeclination)) /
        (cos(latitude.toRadians) * cos(sunDeclination))
    ).toDegrees / 360
```

The issue is in the acos() operation: `Math.acos(-1.091079637558419) == NaN`.
Note that these coordinates are valid: https://goo.gl/maps/t9eDVjFfYT32 :)

**Comment by ticofab** [comment]
As a hint, this library (https://github.com/mikereedell/sunrisesunsetlib-java) fails as well. So there might be something with the coordinates format?

**Comment by oschrenk** [comment]
No. The format is fine. I think the math just fails because the sun is visible the whole day. See https://www.timeanddate.com/sun/@1501821?month=6&year=2017

Without knowing the math, one might be able to fix the error by just identifying if the angle of the sun exceeds a certain threshold. Where is an astronomer if you need one? I will play around a bit.

But this raises also the issue of what you should return as `(sunrise, sunset)` when the sun is always visible, or when the sun is never visible and how to distinguish both edge cases. I'm open for suggestions.

**Comment by oschrenk** [comment]
I am really not sure of the math here, but this _might_ work:
```
    val operand = (sinOfAltitudeSolarDisc - sin(latitude.toRadians) * sin(sunDeclination)) /
        (cos(latitude.toRadians) * cos(sunDeclination))
    if (operand < -1)
      // return no sunset
    else if (operand > 1)
      // return no sunrise
    else {
      val hourAngle = Math.acos(operand).toDegrees / 360
      val sunrise = jTransit - hourAngle
      val sunset = jTransit + hourAngle
      (toZonedDate(sunrise, zoneId), toZonedDate(sunset, zoneId))
    }
```

Not sure what even to return. Maybe just Options, so that?
1. No sunset => `(Some(dateBeginningOfDay), None)`
2. No sunrise => `(None, Some(dateBeginningOfNextDay))`
3. Normal => (Some(dateTime), Some(dateTime))

**Comment by oschrenk** [comment]
See https://github.com/oschrenk/sunrise.scala/tree/polar-circles for the implementation of the idea above.

I will have to figure out where to deploy SNAPSHOTS

**Comment by ticofab** [comment]
Nice work - returning an Option sounds like a good way. Maybe I would even do something like Option[(ZonedDateTime, ZonedDateTime)]. This way you would use None in any case, meaning that the sun is always up or always gone. Or even an Either to specify what the reason is: 
```
Right((ZonedDateTime, ZonedDateTime))
Left(PolarNight) or Left(PolarDay)
```
Your approach is fine too of course :)

**Comment by oschrenk** [comment]
None of those actually makes me really happy. I have to think about it.

Are you able to try out the other code changes in the meantime? I applied for the SNAPSHOT repository via bintray, but that might take a while to be approved I guess.

**Comment by ticofab** [comment]
Another alternative is to return a richer type. Like a `case class SunriseSunsetInfo(...)`, and in it all information you might ever want.

**Comment by oschrenk** [comment]
Fixed in 0.3.0 release

**Comment by ticofab** [comment]
Thanks @oschrenk !
